iam-token-manager/v1.ts

```diff
@@ -0,0 +1,190 @@
+  authHeader?: string;
+  grantType?: string;
+  responseType?: string;
+}
+
+export interface TokenRefreshParams {
+  contentType?: string;
+  authHeader?: string;
+}
+
+export class IamTokenManagerV1 {
+  name: string;
+  serviceVersion: string;
+  protected url: string;
+  protected tokenInfo: IamTokenData;
+  private apikey: string;
+  private userAccessToken: string;
+
+  /**
+   * IAM Token Manager Service
+   *
+   * Retreives, stores, and refreshes IAM tokens.
+   *
+   * @param {Object} options
+   * @param {String} options.iam_apikey
+   * @param {String} options.access_token
+   * @param {String} options.iam_url - url of the iam api to retrieve tokens from
+   * @constructor
+   */
+  constructor(options: Options) {
+    this.url = options.iam_url || 'https://iam.ng.bluemix.net/identity/token';
+    this.tokenInfo = {} as IamTokenData;
+    if (options.iam_apikey) {
+      this.apikey = options.iam_apikey;
+    }
+    if (options.access_token) {
+      this.userAccessToken = options.access_token;
+    }
+  }
+
+  /**
+   * This function sends an access token back through a callback. The source of the token
+   * is determined by the following logic:
+   * 1. If user provides their own managed access token, assume it is valid and send it
+   * 2. If this class is managing tokens and does not yet have one, make a request for one
+   * 3. If this class is managing tokens and the token has expired, refresh it
+   * 4. If this class is managing tokens and has a valid token stored, send it
+   */
+  public getManagedToken(cb: Function) {
+    if (this.userAccessToken) {
+      // 1. use user-managed token
+      return cb(this.userAccessToken);
+    } else if (this.needToRequestToken()) {
+      // 2. request an initial token
+      this.requestToken({}, tokenResponse => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(this.tokenInfo.access_token);
+      });
+    } else if (this.tokenIsExpired()) {
+      // 3. refresh a token
+      this.refreshToken({}, tokenResponse => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(this.tokenInfo.access_token);
+      });
+    } else {
+      // 4. use valid managed token
+      return cb(this.tokenInfo.access_token);
+    }
+  }
+
+  public requestToken(params: TokenRequestParams, cb: Function): void {
+    const contentType = params.contentType ||
+      'application/x-www-form-urlencoded';
+
+    const grantType = params.grantType ||
+      'urn:ibm:params:oauth:grant-type:apikey';
+
+    const authorization = params.authHeader || 'Basic Yng6Yng=';
+    const responseType = params.responseType || 'cloud_iam';
+
+    const requestOptions = {
+      url: this.url,
+      headers: {
+        'Content-type': contentType,
+        Authorization: authorization
+      },
+      form: {
+        grant_type: grantType,
+        apikey: this.apikey,
+        response_type: responseType
+      }
+    };
+    request.post(requestOptions, (err, res, body) => {
```

Comment:
Can we make this method similar to the methods in our services where we build an object and call `requestwrapper`?

Reply:
Yes, I'll look into that